File: NeetCode/08-Tries/ImplementTrie.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;
class Trie {
private:
    struct TrieNode {
        TrieNode* children[26];
        bool isEnd;
        TrieNode() {
            isEnd = false;
            for (int i = 0; i < 26; i++) {
                children[i] = nullptr;
            }
        }
    };
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* curr = root;
        for(int i = 0; i < word.size();i++){
            char c = word[i];
            int index = c - 'a';
            if(curr->children[index] == nullptr){
               curr->children[index] = new TrieNode();
            }
            curr = curr->children[index];
        }
        curr->isEnd = true;
    }
    
    bool search(string word) {
        TrieNode* curr = root; 
       
       for(int i = 0; i < word.size();i++){
            char c = word[i];
            int index = c - 'a';
            if(curr->children[index] == nullptr){
                return false;
            }
            curr = curr->children[index];
       }
       return curr->isEnd;
    }
    
    bool startsWith(string prefix) {
        TrieNode* curr = root;
        
        for(int i = 0; i < prefix.size();i++){
            char c = prefix[i];
            int index = c - 'a';
            if(curr->children[index] == nullptr){
                return false;
            }
            curr = curr->children[index];
        }
        return true;
    }  
};

File: NeetCode/08-Tries/ImplementTrie.java
public class ImplementTrie{
    class TrieNode{
    TrieNode[] children = new TrieNode[26];
    boolean isEnd = false;
    }
    TrieNode root;
    public ImplementTrie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode curr = root;
        for(int i = 0; i < word.length();i++){
            char c = word.charAt(i);
            int index = c - 'a';
            if(curr.children[index] == null){
               curr.children[index] = new TrieNode();
            }
            curr = curr.children[index];
        }
        curr.isEnd = true;
    }
    
    public boolean search(String word) {
       TrieNode curr = root; 
       
       for(int i = 0; i < word.length();i++){
            char c = word.charAt(i);
            int index = c - 'a';
            if(curr.children[index] == null){
                return false;
            }
            curr = curr.children[index];
       }
       return curr.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        
        for(int i = 0; i < prefix.length();i++){
            char c = prefix.charAt(i);
            int index = c - 'a';
            if(curr.children[index] == null){
                return false;
            }
            curr = curr.children[index];
        }
        return true;
    }
}
